keps/sig-storage/3141-prevent-volume-mode-conversion/README.md

```diff
@@ -0,0 +1,724 @@
+   specified in the spec.
+   2. `VolumeSnapshotContent` is created by the `snapshot-controller` in response to (i).
+   3. `snapshot-controller` populates the `Spec` of the given `VolumeSnapshotContent`.
+      1. With this change, the controller will fetch the `Spec.PersistentVolumeMode`
+      of the `PV` and add that to newly introduced `Spec.SourceVolumeMode` field of
+      the VolumeSnapshotContent to be created.
+2. Static Provisioning
+   `VolumeSnapshotContent` is created by the admin. With this change, the admin will be 
+expected to fill the `Spec.SourceVolumeMode` field appropriately. If left nil, `Unknown`
+mode will be assumed to preserve existing behavior.
+
+### Changes to external-provisioner
+
+This design leverages the access rights of a user on `VolumeSnapshotContents` to
+determine whether the volume mode can be modified when a `PVC` is being created
+with a `VolumeSnapshot` as the source.
+The volume mode can be altered if the requesting user has `Update` and `Patch` rights
+on `VolumeSnapshotContents` (which is a cluster scoped resource).
+The control flow for creating a `PVC` from a `VolumeSnapshot` will look like below:
+
+1. A user attempts to create a `PVC` from a `VolumeSnapshot` by specifying the
+`Spec.DataSource` parameter of the `PVC` YAML.
+2. `external-provisioner` receives a callback to dynamically create the volume.
+As part of the preprocessing steps, it will:
+   1. Get the `Spec.SourceVolumeMode` of the `VolumeSnapshotContent`.
+      1. If `Spec.SourceVolumeMode` doesn't exist or is nil, then continue with
+      volume provisioning to preserve existing behavior.
+   2. Get the `Spec.VolumeMode` of the `PVC` being created.
+   If they do not match:
+      1. Get all annotations on the `VolumeSnapshotContent` and verify if 
+      `snapshot.storage.kubernetes.io/allowVolumeModeChange: true` exists.
+      If it does not exist, block volume provisioning by returning an error.
+4. In all other cases, let volume provisioning continue.
+
+NOTE: `external-provisioner` maintains a reference to `PVC` and `VolumeSnapshotContent`
+during volume creation. This proposal leverages those references to make additional
+decisions.
+
+### Test Plan
+
+E2E tests will be added for this design, that attempt to restore a volume with
+and without requisite privileges. 
+
+#### Unit tests
+
+- With feature gate disabled:
+  - attempt to convert volume mode when creating a `PVC`
+  from a `VolumeSnapshot`.
+- With feature gate enabled, attempt to convert volume mode when creating a `PVC`
+from a `VolumeSnapshot`:
+  - With `Spec.SourceVolumeMode` populated and `snapshot.storage.kubernetes.io/allowVolumeModeChange: true`
+  annotation present.
+  - With `Spec.SourceVolumeMode` populated but no `snapshot.storage.kubernetes.io/allowVolumeModeChange: true`
+  annotation.
+  - With `Spec.SourceVolumeMode` set to `nil`.
+
+#### E2E tests
+
+- Attempt to convert volume mode when creating a `PVC`from a `VolumeSnapshot`:
```

Comment:
Is this with feature flag disabled or enabled or both?

Reply:
I was thinking e2e tests can only be run with the feature flag enabled.. is it possible to tweak the feature flag in e2e tests too? IIUC e2e tests will be added directly to k/k

Reply:
That's fine. Can you clarify here feature flag will be enabled for these e2e tests?

Reply:
Done